control/controlclient/direct.go

```diff
@@ -1204,6 +1218,40 @@ func sleepAsRequested(ctx context.Context, logf logger.Logf, timeoutReset chan<-
 	}
 }
 
+// getNoiseClient returns the noise client, creating one if one doesn't exist.
+func (c *Direct) getNoiseClient(ctx context.Context) (*noiseClient, error) {
+	c.mu.Lock()
+	su := c.serverURL
+	np := c.noiseClient
+	sk := c.serverNoiseKey
+	c.mu.Unlock()
+	if sk.IsZero() {
+		return nil, errors.New("zero serverNoiseKey")
+	}
+	if np != nil {
+		return np, nil
+	}
+	k, err := c.getMachinePrivKey()
+	if err != nil {
+		return nil, err
+	}
+
+	np, err = newNoiseClient(ctx, k, sk, su)
+	if err != nil {
+		return nil, err
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
```

Comment:
use a SingleFlight to avoid having these races?

Reply:
I don't think `singleflight` helps us much here. We need to be able to create multiple connections (in case of trunkd restarts, local client errors etc), and we need the mutex anyways for cleaning up/closing.

Reply:
oops, you're right this can be simplified by singleflight. I thought this was in `noise.go`.